### PR TITLE
Log when an insecure Https -> Http redirect is blocked

### PR DESCRIPTION
**Pull request by rmkerr**
.NET Core does not allow insecure redirects from https to http, but .NET Framework does. Developers have had some trouble diagnosing the issue when they run into this difference. This change adds logging to help developers track down what is going on.

Fixes: #23813 (was https://github.com/dotnet/corefx/issues/24577)

### CODE REVIEW

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: nit: add space after 'if'

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs]: nit: add space after 'if'

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: It would be nice if we had EventSourceTrace style tracing available in WinHttpHandler (similar to CurlHandler, etc.), but that is probably beyond the scope of this PR.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: nit: bracing style should match rest of file, i.e. Allman-style

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs
Comment (davidsh) [reply src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs]: nit: bracing style should match rest of file, i.e. Allman-style

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (caesar1995) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Should we consider status code 308: Permanent Redirect here?

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs]: Interestingly we could use this same check to avoid doing some of the above work if we're not going to redirect anyway.  But it's probably not something worth optimizing for.

**Comment by stephentoub** [comment]
Can you add similar tracing to SocketsHttpHandler?  Probably would go here:
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticateAndRedirectHandler.cs#L135

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (davidsh) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Maybe it would be less code/higher perf to just check for any 3xx code, i.e. do some simple math operation than listing all these out.

Comment (rmkerr) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: I considered that but was concerned about `304 Not Modified` responses, which have a different behavior.

**Comment by rmkerr** [comment]
Thanks for the feedback! I'll send out another iteration with the minor changes and with SocketsHttpHandler included.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (danmosemsft) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: Are the urls in the payload or should they be included in the trace message?

Comment (rmkerr) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: They're not in the payload. so I'll add it to the message. What do you think of wording it something like this?
```
$"Insecure https to http redirect from {RequestUri} to {DestinationUri} blocked."
```

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (rmkerr) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: This whole block is pretty ugly, but in most cases it will make fewer checks than it would before. It also covers the case of `HttpStatusCode.PermanentRedirect`, which it did not previously.

**Comment by rmkerr** [comment]
I've addressed review comments and added logging to `SocketsHttpHandler`. While doing that, I did notice that `SocketsHttpHandler` has extremely limited logging compared to `WinHttpHandler` and 'CurlHandler`. I don't think that is within the scope of this PR, but I think that it is worth looking into later.

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs]: This is unfortunately a relatively expensive operation.  This is translated by the compiler into:
```C#
EventSourceTrace(string.Format("Insecure https to http redirect from {0} to {1} blocked.", new object[] { _requestMessage.RequestUri, newUri }));
```
so even if tracing is disabled, we still end up allocating that object[], and we still end up doing the string format.

It should instead do something like:
https://github.com/rmkerr/corefx/blob/c1a2eddf4ee509b1e591f8e63ad7429625e42b20/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs#L190
where the argument is passed in as an argument, and all of the work is deferred until in the EventSourceTrace method it checks for whether tracing is enabled.  I don't recall for sure, but I don't think there's currently an EventSourceTrace overload that takes two arguments to be substituted; assuming not, you could just use a value tuple, e.g.
```C#
EventSourceTrace("Insecure https to http redirect: {0}", (_requestMessage.RequestUri, newUri));
```

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: See my comment on the CurlHandler change.  There are similar costs here.  More, actually, as this will create an extra string for concatenating the const string with the result of the string interpolation.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticateAndRedirectHandler.cs
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticateAndRedirectHandler.cs]: Unlike the other two, this one is fine, as it's guarded by the NetEventSource.IsEnabled check.  The costs will only be incurred when tracing is enabled.

**Comment by stephentoub** [comment]
> While doing that, I did notice that SocketsHttpHandler has extremely limited logging compared to WinHttpHandler and CurlHandler.

That's true in that auth/redirect file.  It could certainly stand to be improved.  That said, @geoffkizer is currently doing some refactoring here, so I'd recommend holding off on significant changes there until his changes land.

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs
Comment (rmkerr) [reply src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs]: I'm used to thinking of that kind of string operation as "free", but obviously that isn't the case with format strings. I'll make the change, and the cost of format strings is something I'll keep in mind!

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (rmkerr) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: This should be mostly resolved. I can save the cost of the call to format, but do still incur the cost of casting both URIs to strings. I could fix that, but as far as I can tell I would have to modify WinHttpTraceHelper to do that.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: If you need to wrap this so that this doesn't cause extra work if tracing isn't enabled, then wrap this in a if (WinHttpTraceHelper.IsEnabled) etc.

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs
Comment (pjanotti) [inline src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs]: Remove the interpolation ```$```

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs
Comment (rmkerr) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs]: I'll make that change. Even though it means making that truly horrible if-statement even worse.

File: src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs
Comment (rmkerr) [reply src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs]: I missed that you made this comment! I'll try to address that.

**Review by pjanotti — APPROVED** [review]
:shipit:

**Comment by rmkerr** [comment]
CI is green. @stephentoub @davidsh any additional comments, or are you okay with me going ahead and merging this?

**Comment by stephentoub** [comment]
LGTM.  FYI this is going to conflict with @geoffkizer's change at https://github.com/dotnet/corefx/pull/27266.

**Review by davidsh — APPROVED** [review]

**Comment by rmkerr** [comment]
Hey @geoffkizer! Would you prefer if I held off on merging this until your PR is through? I think it should be easy to reconcile the changes either way.

**Pull request merged**
